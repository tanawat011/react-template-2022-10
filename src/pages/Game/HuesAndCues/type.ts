/* istanbul ignore file */
import type { DocumentData, QueryDocumentSnapshot } from 'firebase/firestore'

export type Player = {
  id: string
  name: string
}

export type Room = {
  id: string
  password: string
  hintChoice: string[]
  hintSelected: string
  isStarted: boolean
  isSubmitResult: boolean
  totalRound: number
}

export type RawRoomPlayer = {
  refPlayer: QueryDocumentSnapshot<DocumentData>
  color: string
  score: number
  allSelected: string[]
  isOwner: boolean
  isTurn: boolean
  isHinter: boolean
  totalTurn: number
  seq: number
}

export type RoomPlayer = Omit<RawRoomPlayer, 'refPlayer'> & {
  player: Player
}
