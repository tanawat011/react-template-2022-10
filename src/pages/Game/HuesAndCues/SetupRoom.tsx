import type { Player } from './_type'

import type { FormEvent, MouseEvent } from 'react'
import { useEffect, useState } from 'react'

import { useNavigate } from 'react-router-dom'

import { Button } from 'components/Button'
import { useSessionStorage } from 'hooks'

import { SESSION } from './_constants'
import {
  getPlayerProfile,
  setPlayerInTheRoom,
  toRoom,
  toSetDisplayRoom,
  toSetupRoom,
} from './services'

export const SetupRoom = () => {
  const navigate = useNavigate()
  const [sessionId] = useSessionStorage(SESSION.ID)
  const [roomId, setRoomId] = useSessionStorage(SESSION.ROOM_ID)

  const [player, setPlayer] = useState<Player>()

  useEffect(() => {
    if (!sessionId) {
      toSetDisplayRoom(navigate)

      return
    }

    if (roomId) {
      toRoom(navigate, roomId)

      return
    }

    fetchPlayer()
  }, [])

  const fetchPlayer = async () => {
    const _player = await getPlayerProfile(sessionId)

    setPlayer(_player)
  }

  const handleCreatePublicRoom = async (e: MouseEvent) => {
    e.preventDefault()

    if (player) {
      const uid = sessionId
      const _roomId = sessionId

      try {
        // * Create room and add player to the room
        await setPlayerInTheRoom(_roomId, {
          uid,
          name: player.name,
          score: 0,
          color: '',
          cells: [],
          number: 1,
          hinter: true, // * The first player is the hinter
          owner: true, // * Owner is the first player to join the room
        })

        setRoomId(_roomId)

        toRoom(navigate, _roomId)
      } catch (error) {
        // eslint-disable-next-line no-console
        console.log('error', error)
      }
    }
  }

  const handleCreatePrivateRoom = async (e: MouseEvent) => {
    e.preventDefault()

    try {
      toSetupRoom(navigate)
    } catch (error) {
      // eslint-disable-next-line no-console
      console.log('error', error)
    }
  }

  const handlerSubmitJoinPrivateRoom = async (e: FormEvent) => {
    e.preventDefault()

    // eslint-disable-next-line no-console
    console.log('handlerSubmitJoinPrivateRoom')
  }

  return (
    <div>
      <Button type='button' onClick={handleCreatePublicRoom}>
        Create Public Room
      </Button>

      <Button type='button' onClick={handleCreatePrivateRoom}>
        Create Private Room
      </Button>

      <form onSubmit={handlerSubmitJoinPrivateRoom}>
        <label id='code'>Code</label>
        <input id='code' />
        <Button type='submit'>Join Private Room</Button>
      </form>
    </div>
  )
}
