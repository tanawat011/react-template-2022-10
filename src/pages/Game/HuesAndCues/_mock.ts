import type { RoomPlayer } from './_type'

const PLAYER_BASE: RoomPlayer = {
  cells: [],
  uid: '1',
  name: 'Player 1',
  score: 0,
  color: '',
  number: 1,
}

const HINTER: RoomPlayer = {
  uid: 'p1',
  name: 'p1',
  number: 1,
  color: 'bg-[#0758a5]',
  score: 0,
  hinter: true,
  cells: [],
}

const HINTER_HAS_CHOICE: RoomPlayer = {
  ...HINTER,
  choice: ['A1', 'A2', 'A3', 'A4'],
}

const HINTER_HAS_CHOICE_SELECT_RESULT: RoomPlayer = {
  ...HINTER_HAS_CHOICE,
  result: 'A1',
}

const HINTER_HAS_CHOICE_SELECT_RESULT_SUBMIT_RESULT: RoomPlayer = {
  ...HINTER_HAS_CHOICE_SELECT_RESULT,
  isSubmitResult: true,
}

const HINTER_IS_OWNER: RoomPlayer = {
  ...HINTER,
  owner: true,
}

const HINTER_IS_OWNER_HAS_CHOICE: RoomPlayer = {
  ...HINTER_HAS_CHOICE,
  owner: true,
}

const HINTER_IS_OWNER_HAS_CHOICE_SELECT_RESULT: RoomPlayer = {
  ...HINTER_HAS_CHOICE_SELECT_RESULT,
  owner: true,
}

const HINTER_IS_OWNER_HAS_CHOICE_SELECT_RESULT_SUBMIT_RESULT: RoomPlayer = {
  ...HINTER_HAS_CHOICE_SELECT_RESULT_SUBMIT_RESULT,
  owner: true,
}

const PLAYER: RoomPlayer = {
  uid: 'P2',
  name: 'p2',
  number: 2,
  color: 'bg-[#117822]',
  score: 0,
  cells: [],
}

const PLAYER_MY_TURN: RoomPlayer = {
  ...PLAYER,
  isMyTurn: true,
}

const PLAYER_MY_TURN_SELECT_CELL: RoomPlayer = {
  ...PLAYER_MY_TURN,
  cells: ['A1'],
}

const result = 'F4'
const choice = ['F4', 'A9', 'P27', 'E15']

export const MOCK: {
  ONE_PLAYER: RoomPlayer[]
  TWO_PLAYER: RoomPlayer[]
  TWO_PLAYER_LAST_IS_HINTER: RoomPlayer[]
  THREE_PLAYER: RoomPlayer[]
  THREE_PLAYER_SECOND_IS_HINTER: RoomPlayer[]
  THREE_PLAYER_LAST_IS_HINTER: RoomPlayer[]
  THREE_PLAYER_LAST_IS_TURN: RoomPlayer[]
  HINTER_NO_CHOICE: RoomPlayer[]
  HINTER_NO_SELECT_CHOICE: RoomPlayer[]
  HINTER_SELECTED_CHOICE: RoomPlayer[]
  PLAYER_NO_COLOR: RoomPlayer[]
  PLAYER_HAS_COLOR: RoomPlayer[]
  PLAYER_HAS_COLOR_START_GAME: RoomPlayer[]
  PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT: RoomPlayer[]
} = {
  ONE_PLAYER: [
    {
      ...PLAYER_BASE,
      cells: ['D17'],
      number: 1,
      hinter: true,
      isMyTurn: true,
      color: 'bg-[#0758a5]',
    },
  ],
  TWO_PLAYER: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
  ],
  TWO_PLAYER_LAST_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      hinter: true,
    },
  ],
  THREE_PLAYER: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      result,
      choice,
      owner: true,
      score: 1,
    },
    {
      ...PLAYER_BASE,
      uid: '2',
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      uid: '3',
      number: 3,
    },
  ],
  THREE_PLAYER_LAST_IS_TURN: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      result,
      choice,
    },
    {
      ...PLAYER_BASE,
      number: 2,
    },
    {
      ...PLAYER_BASE,
      number: 3,
      isMyTurn: true,
    },
  ],
  THREE_PLAYER_SECOND_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  THREE_PLAYER_LAST_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
      hinter: true,
    },
  ],
  HINTER_NO_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  HINTER_NO_SELECT_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      choice,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  HINTER_SELECTED_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      choice,
      result,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_NO_COLOR: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_HAS_COLOR: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
      color: 'bg-[#117822]',
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_HAS_COLOR_START_GAME: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
      isStartGame: true,
      result: 'F4',
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
      cells: ['F4'],
    },
    {
      ...PLAYER_BASE,
      number: 3,
      cells: ['F5', 'F6'],
    },
  ],
  PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
      isStartGame: true,
      isSubmitResult: true,
      countRound: 1,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
}
