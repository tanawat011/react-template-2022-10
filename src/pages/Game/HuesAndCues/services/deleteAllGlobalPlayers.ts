import type { RoomPlayer } from '../type'

import { deleteDocument } from 'helpers/firebase'

import { FIRESTORE_PATH } from '../constants'

export const deleteAllGlobalPlayers = async (players: RoomPlayer[]) => {
  const playerPath = FIRESTORE_PATH.DOC_PLAYER

  await Promise.all(
    players.map(async ({ player: { id } }) => {
      const path = playerPath + `/${id}`

      await deleteDocument(path)
    }),
  )
}
