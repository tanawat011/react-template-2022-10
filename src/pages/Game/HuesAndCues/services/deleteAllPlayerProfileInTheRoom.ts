import type { RoomPlayer } from '../_type'

import { getFirestore } from 'firebase/firestore'

import { deleteDocument } from 'helpers/firebase'

import { FIRESTORE_PATH } from '../_constants'

export const deleteAllPlayerProfileInTheRoom = async (players: RoomPlayer[]) => {
  const db = getFirestore()

  await Promise.all(
    players.map(async ({ uid }) => {
      const collectionPath = `${FIRESTORE_PATH.PLAYERS_COLLECTION}/${uid}`

      await deleteDocument(db, collectionPath)
    }),
  )
}
