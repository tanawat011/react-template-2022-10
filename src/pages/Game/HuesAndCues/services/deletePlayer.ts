import { FIRESTORE_PATH } from 'constants/gameHuesAndCues'
import { deleteCollection } from 'helpers/firebase'

export const deletePlayer = async (playerId: string) => {
  const path = FIRESTORE_PATH.DOC_PLAYER + `/${playerId}`

  await deleteCollection(path)
}
