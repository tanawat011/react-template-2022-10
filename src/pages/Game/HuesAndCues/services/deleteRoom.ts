import { getFirestore } from 'firebase/firestore'

import { deleteCollection } from 'helpers/firebase'

import { FIRESTORE_PATH } from '../_constants'

export const deleteRoom = async (roomId: string) => {
  const db = getFirestore()
  const collectionPath = `${FIRESTORE_PATH.ROOMS_DOCUMENT}/${roomId}`

  await deleteCollection(db, collectionPath)
}
