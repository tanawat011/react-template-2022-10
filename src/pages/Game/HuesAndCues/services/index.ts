export * from './deleteAllPlayerProfileInTheRoom'
export * from './deleteRoom'
export * from './getAllPlayersInTheRoom'
export * from './getPlayerInTheRoom'
export * from './getPlayerProfile'
export * from './setPlayerProfile'
export * from './setPlayerInTheRoom'
export * from './subscribeRoom'

export * from './navigate'
