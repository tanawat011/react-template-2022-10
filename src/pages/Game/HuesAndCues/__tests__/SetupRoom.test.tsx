import type { RenderResult } from '@testing-library/react'

import { act, fireEvent, screen, waitFor } from '@testing-library/react'
import { mocked } from 'jest-mock'

import { renderWithProviders } from 'helpers/test'

import { SetupRoom } from '../SetupRoom'
import { MOCK } from '../_mock'
import { getPlayerProfile, setPlayerInTheRoom } from '../services'

jest.mock('components/Button', () => ({
  Button: (props: React.HTMLAttributes<never>) => {
    return (
      <div {...props} data-testid={'button'}>
        {props.children}
      </div>
    )
  },
}))

jest.mock('../services', () => ({
  getPlayerProfile: jest.fn(),
  setPlayerInTheRoom: jest.fn(),
  toRoom: jest.fn(),
  toSetDisplayRoom: jest.fn(),
  toSetupRoom: jest.fn(),
}))

describe('<SetupRoom />', () => {
  afterEach(() => {
    sessionStorage.removeItem('sessionId')
    sessionStorage.removeItem('roomId')
  })

  test('match snapshot', () => {
    sessionStorage.setItem('sessionId', 'session-id-x')

    const { asFragment } = renderWithProviders(<SetupRoom />)

    expect(asFragment()).toMatchSnapshot()
  })

  test('no session id', () => {
    const { asFragment } = renderWithProviders(<SetupRoom />)

    expect(asFragment()).toBeTruthy()
  })

  test('has session id and room id', () => {
    sessionStorage.setItem('sessionId', 'session-id-x')
    sessionStorage.setItem('roomId', 'room-id-x')

    const { asFragment } = renderWithProviders(<SetupRoom />)

    expect(asFragment()).toBeTruthy()
  })

  test('create public room', async () => {
    mocked(getPlayerProfile).mockResolvedValue(MOCK.ONE_PLAYER[0])
    sessionStorage.setItem('sessionId', 'session-id-x')
    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(<SetupRoom />)
    })

    const button = screen.getAllByTestId('button')[0]
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('create public room, but no players', async () => {
    sessionStorage.setItem('sessionId', 'session-id-x')
    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(<SetupRoom />)
    })

    const button = screen.getAllByTestId('button')[0]
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('create public room, but error', async () => {
    mocked(getPlayerProfile).mockResolvedValue(MOCK.ONE_PLAYER[0])
    mocked(setPlayerInTheRoom).mockRejectedValue('error')
    sessionStorage.setItem('sessionId', 'session-id-x')
    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(<SetupRoom />)
    })

    const button = screen.getAllByTestId('button')[0]
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('create private room', async () => {
    sessionStorage.setItem('sessionId', 'session-id-x')
    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(<SetupRoom />)
    })

    const button = screen.getAllByTestId('button')[1]
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('join room', async () => {
    sessionStorage.setItem('sessionId', 'session-id-x')
    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(<SetupRoom />)
    })

    const button = screen.getAllByTestId('button')[2]
    fireEvent.submit(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })
})
