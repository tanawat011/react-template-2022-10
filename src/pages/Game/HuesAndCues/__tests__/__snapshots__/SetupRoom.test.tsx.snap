// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SetupRoom /> match snapshot 1`] = `
<DocumentFragment>
  <div>
    <button
      data-testid="button"
      type="button"
    >
      Create Public Room
    </button>
    <button
      data-testid="button"
      type="button"
    >
      Create Private Room
    </button>
    <form>
      <label
        id="code"
      >
        Code
      </label>
      <input
        id="code"
      />
      <button
        data-testid="button"
        type="submit"
      >
        Join Private Room
      </button>
    </form>
  </div>
</DocumentFragment>
`;
